design/restful-interface

```diff
@@ -0,0 +1,390 @@
+
+We will add ${definitionType} in the path to differentiate if it's a workload/trait/scope definition.
+It can be either "workloadDefinition","traitDefinition", or "scopeDefinition" as they are three different
+types of resources.
+ 
+### Create 
+
+**URL** : `/api/${definitionType}`
+
+**Method** : `POST`
+
+**Query Parameter** : `None`
+
+**Body** :
+
+```go
+
+type oamDefinition struct {
+	definitionName string `json:"definitionName"`
+    definition runtime.RawExtension `json:"definition"` // the real definition data
+}
+```
+
+### Update 
+
+**URL** : `/api/${definitionType}/${definitionName}`
+
+**Method** : `PUT`
+
+**Query Parameter** : `updateMode` 
+
+**Body** :
+
+```go
+
+type oamDefinition struct {
+    definition runtime.RawExtension `json:"definition"` // the definition data
+}
+```
+
+**Responses Body** : None
+
+### GET
+
+**URL** : `/api/${definitionType}/${definitionName}`
+
+**Method** : `Get`
+
+**Query Parameter** : `qureyMode` 
+
+**Body** : None
+
+**Responses Body** : 
+```go
+type definition struct {
+    definition runtime.RawExtension `json:"definition"` // either the full definition or the cueTempalte
+}
+```
+
+### List
+
+**URL** : `/api/${definitionType}`
+
+**Method** : `Get`
+
+**Query Parameter** : `None`
+
+**Body** : None
+
+**Responses Body** : 
+```go
+type definitionName struct {
+    definitionNames string[] `json:"definitionNames"` // the definition name list, no next Token
+}
+```
+
+### Delete
+
+**URL** : `/api/${definitionType}/${definitionName}`
+
+**Method** : `Delete`
+
+**Query Parameter** : `dryrun` 
+
+**Body** : None
+
+**Responses Body** : None 
+
+
+## Repo related API
+These API operate on definition files stored in a git repo. It could be a git repo on the local file system. 
+
+They all have a `repoURL` query parameter that stores the git repo's URL so that the server is stateless. 
+ 
+### Update 
+
+**URL** : `/api/category/${definitionType}/${definitionName}`
+
+**Method** : `PUT`
+
+**Query Parameter** : `repoURL`, `updateMode` 
+
+**Body** :
+
+```go
+
+type oamDefinition struct {
+    definition runtime.RawExtension `json:"definition"` // the definition data
+}
+```
+
+**Responses Body** : None
+
+### GET
+
+**URL** : `/api/category/${definitionType}/${definitionName}`
+
+**Method** : `Get`
+
+**Query Parameter** : `repoURL`, `qureyMode` 
+
+**Body** : None
+
+**Responses Body** : 
+```go
+type definition struct {
+    definition runtime.RawExtension `json:"definition"` // either the full definition or the cueTempalte
+}
+```
+
+### List
+
+**URL** : `/api/category/${definitionType}`
+
+**Method** : `Get`
+
+**Query Parameter** : `repoURL`, `None`
+
+**Body** : None
+
+**Responses Body** : 
+```go
+type definitionName struct {
+    definitionNames string[] `json:"definitionNames"` // the definition name list, no next Token
+}
+```
+
+## Version API
+This is debatable, we are not sure if this is really needed.
+
+### GET
+
+**URL** : `/api/version/${version}`
```

Comment:
```suggestion
**URL** : `/api/version`
```

Reply:
normally get has an instance name there, I am thinking of using $latest as the default. Without it, the version itself is a singleton resource. I am